### ISSUE TITLE
Watch mode doesn't handle filesystem changes

### ISSUE THREAD

**Issue by eyelidlessness** [rsms/estrella#29]
### Steps to reproduce

1. Configure a `build` with a list of entries (e.g. by glob)
2. Run the build with `-w`/`--watch`
3. Rename one of the entries

### Expected

Well, it's not clear what's expected because the API doesn't support glob entries or a way to provide them dynamically. But ideally there would be a way to react to filesystem changes, somehow.

### Observed

`error: Could not resolve [... previous path to the file ...]`

### Mitigation

For anyone looking to work around this limitation, I've used a flag called `once` (so watch mode isn't automatically enabled), run an initial build, then call the `watch` function to run subsequent builds, which retrieves a new list of entries on each change. (This is probably not the *most* efficient, as the `watch` function provides information about what changed, but I've honestly lost too much time today to worry about that right now.)

**Comment by rsms** [comment]
To be clear: This is only a limitation with `entryPoint` files.

Solving this is actually very hard since there's natural ambiguity. When a file is renamed, for example foo.js → bar.js, this is what most OSes will report:

- foo.js deleted
- bar.js created

What makes this hard:

- order is not guaranteed; some programs may write bar.js first and unlink foo.js afterward
- usually bar.js will receive a different inode than foo.js so you can't rely on that
- performing content comparison has a race condition

The least unreliable method is probably content comparison.
You'd do something like this:
1. checksum foo.js whenever it changes
2. when it is deleted, wait for a short amount of time to allow writes of the new file to finish
3. scan all files in the directory tree and hash their contents (this can be very slow if there are a lot of files in the source tree)
4. if another file has the same content as the previously-existing foo.js and that file appeared within a short time window of foo.js disappearing, you make a (risky) assumption that "foo.js is now called bar.js".

Clearly some issue with this approach. For example, the user may use an editor that allows changing the name of the file while also saving changes, which means that the checksum of the "new" file (bar.js) will not match that of the "old" fole foo.js. Another issue is that with step 4 — it is not unheard of to have two files with the same content but different names (maybe same basename but in different directories.) If you were to write a file like that within your "peek" time window in step 4, you might accidentally map the rename to the wrong file.

However on some systems we can get reliable filesystem events. For example, on macOS we can listen for events directly from fsevents, skipping Chokidar's interpretation (which is what Estrella uses and it does not support "rename"). macOS kqueue will issue two consecutive (second immediate) "moved" events when a file is renamed. We can use this to support rename tracking on _some_ platforms.

**Comment by rsms** [comment]
I've started some initial work on branch https://github.com/rsms/estrella/compare/watch-rename
You can try it by checking out the source and
1. run `./build.js -g`
2. run `(cd test/watch && ESTRELLA_TEST_VERBOSE=1 node test-rename.js)`

You should see this:

```
$ (cd test/watch && ESTRELLA_TEST_VERBOSE=1 node test-rename.js)
Wrote tmp-out.js (16B, 5.94ms)
Watching files for changes...
~~~~~~~~~~~ step 1 ~~~~~~~~~~~
moving file 'tmp-in.ts' -> 'tmp-in-renamed.ts'
1 file changed: tmp-in.ts
Wrote tmp-out.js (16B, 2.56ms)
~~~~~~~~~~~ step 2 ~~~~~~~~~~~
writing to new file 'tmp-in-renamed.ts'
1 file changed: tmp-in-renamed.ts
Wrote tmp-out.js (16B, 1.84ms)
~~~~~~~~~~~ step 3 ~~~~~~~~~~~
buildProcess ended { ok: true }
PASS
```

For detailed information, turn on debug logging: `(cd test/watch && ESTRELLA_TEST_VERBOSE=1 node test-rename.js -estrella-debug)`

**Comment by eyelidlessness** [comment]
I appreciate all of the detail in your response! And share your concerns about the various ambiguities that arise. I can’t look at your branch atm but did want to give this an initial response and some thoughts on the challenges.

- There’s already facilities for addressing race conditions in a debounce, it could be helpful here.
- Git will often be available and can help provide rename+edit insights.
- The whole build process is so fast that just invalidating the whole cache when the tree changes is probably a reasonable first stab at it.
- Worth a note: renaming files is definitely a common task but adding new ones was where I ran into trouble. My goal wasn’t to get super efficient about refactoring, just to see changes _at all_ when I was building new content/functionality, without restarting the process.

**Comment by rsms** [comment]
> There’s already facilities for addressing race conditions in a debounce, it could be helpful here.

Could you point me in the right direction? Perhaps a ref implementation. I'm not sure what you mean with "debounce" in this context.

> Git will often be available and can help provide rename+edit insights

Agreed, however Estrella can't rely on git so I'd leave this for the author of their build.js script to implement. I.e. a simple exec with child_process can be used to trigger rebuilds and reconfigure a build script.

> The whole build process is so fast that just invalidating the whole cache when the tree changes is probably a reasonable first stab at it.

What does "invalidating the whole cache" mean? Estrella doesn't cache information about the file system. It simply asks the OS to inform it about changes to the sources files which esbuild reports as "used for compilation." You could skip the `-watch` flag and instead use estrella.watch in your build script to watch your entire source directory and trigger a rebuild whenever it changes.

> [...] adding new ones was where I ran into trouble

Estrella only watches files for changes which are actually used in the compilation. That means that until you import or require a new file, it won't trigger any builds. This is a good thing since it means that Estrella is really fast and only rebuilds when needed. This works in cooperation with esbuild which provides estrella with a list of all source files actually used in a compilation :-)

**Comment by eyelidlessness** [comment]
> Could you point me in the right direction? Perhaps a ref implementation. I'm not sure what you mean with "debounce" in this context.

Sure, [here’s one](https://github.com/lodash/lodash/blob/4.17.15/lodash.js#L10304). (I’m not endorsing lodash, just linking it because I’m on phone and it was easiest for me to find.)

> Agreed, however Estrella can't rely on git so I'd leave this for the author of their build.js script to implement. I.e. a simple exec with child_process can be used to trigger rebuilds and reconfigure a build script.

This seems perfectly reasonable.

> What does "invalidating the whole cache" mean? Estrella doesn't cache information about the file system. It simply asks the OS to inform it about changes to the sources files which esbuild reports as "used for compilation." You could skip the -watch flag and instead use estrella.watch in your build script to watch your entire source directory and trigger a rebuild whenever it changes.

This is essentially what I am doing right now as a work around.

This is probably a good point to make my use case more clear, as I’m not using Estrella to *build* but effectively as a faster `ts-node` for test runs. Like most test runners, the ones I use (currently Ava for personal projects, Uvu for one I contribute to; Uvu doesn’t have a watch mode which is what led me to Estrella in the first place) *also* take a glob and imports matching modules dynamically. So there’s no real entrypoint. There surely could be, but that’s a fair bit of DX friction for a TDD workflow.

For what it’s worth, I’m currently looking into moving to another approach, using `esbuild` transform dynamically at import time. But I wanted to be sure to follow up since I did open this issue, and answering your followup questions might help Estrella serve this use case in the future.

**Comment by rsms** [comment]
> > Could you point me in the right direction? Perhaps a ref implementation. I'm not sure what you mean with "debounce" in this context.
> 
> Sure, [here’s one](https://github.com/lodash/lodash/blob/4.17.15/lodash.js#L10304). (I’m not endorsing lodash, just linking it because I’m on phone and it was easiest for me to find.)

I'm not sure how a generic debounce function would help. Are you imagining using a debounce methodology for waiting for events? I'm sorry, I think you lost me here. The issue with reliability it not about event frequency or timing even, but that of uncertainty; the fact that at between two points in time things might happen that you can not know and so you have to make the assumption that stuff always happens between to measurements or readings.

> This is probably a good point to make my use case more clear, as I’m not using Estrella to _build_ but effectively as a faster `ts-node` for test runs. Like most test runners, the ones I use (currently Ava for personal projects, Uvu for one I contribute to; Uvu doesn’t have a watch mode which is what led me to Estrella in the first place) _also_ take a glob and imports matching modules dynamically. So there’s no real entrypoint. There surely could be, but that’s a fair bit of DX friction for a TDD workflow.

Ah, I see. Thanks for clarifying that. Estrella only handles watching of source files. For anything not part of a compilation source tree you'll have to either use the provided `watch` function (which is a light wrapper around Chokidar, accepting all Chokidar options) or use something else.

> For what it’s worth, I’m currently looking into moving to another approach, using `esbuild` transform dynamically at import time. But I wanted to be sure to follow up since I did open this issue, and answering your followup questions might help Estrella serve this use case in the future.

👍

**Comment by eyelidlessness** [comment]
> I'm not sure how a generic debounce function would help. Are you imagining using a debounce methodology for waiting for events? I'm sorry, I think you lost me here.

It would allow reconciliation of a created file and a deleted file, which was in reality a “moved” or “renamed” file. Denouncing the two events would let you perform a diff (which could use git’s preferred algorithm even without a repo present), to infer a probable path change.

**Comment by rsms** [comment]
Shipped with [release 1.4.0](https://github.com/rsms/estrella/releases/tag/v1.4.0)
